### ISSUE TITLE
Memory issue when using fused_dense kernel with deepspeed

### ISSUE THREAD

**Issue by yongyanrao** [HazyResearch/flash-attention#222]
Hi, we are having a memory issue when using fused_dense kernel with deepspeed.

We originally have deepspeed installed and worked properly, and wanted to use fused_dense kernel. So we did `pip install .` from `flash-attention/csrc/fused_dense_lib`. After the installation, we called `fused_mlp_func()` to use the kernel. What we can see is that
1. If we left the `process_group` to the default value, we can see the GPU memory usage got accumulated through training, and it ran into OOM issue in around 25 iterations.
2. If we used `process_group=torch.distributed.new_group([torch.distributed.get_rank()])`, the training ran into OOM issue in around 57 iterations.

We would like to know how we can avoid the memory leaking.

**Comment by tridao** [comment]
I don't use deepspeed so idk if it messes with things. I've had models running for weeks with `fused_dense` without any memory issue.
To help pin point the issue, can you try these to see if there's still memory issue:
1. Without deepseed.
2. With apex's [fused_dense](https://github.com/NVIDIA/apex/blob/master/apex/fused_dense/fused_dense.py) instead of `fused_dense` from this repo.

**Comment by yongyanrao** [comment]
@tridao Thank you so much for the response. Yes, if we turned off deepspeed, there was no memory issue. We will do some tests with apex's fused_dense, and report here.

**Comment by yongyanrao** [comment]
We are still working on testing apex's fused_dense, and noticed an issue that the output of `_fused_dense_gelu_dense` from https://github.com/NVIDIA/apex/blob/master/apex/fused_dense/fused_dense.py#L59 does not have the sequence length dimension, e.g., the tensors in
```
out = _fused_dense_gelu_dense(input, self.dense1.weight, self.dense1.bias, self.dense2.weight, self.dense2.bias)
```
have shapes
- out: torch.Size([8, 768]) **We are expecting torch.Size([8, 1024, 768])**
- input: torch.Size([8, 1024, 768])
- self.dense1.weight: torch.Size([3072, 768])
- self.dense1.bias: torch.Size([3072])
- self.dense2.weight: torch.Size([768, 3072])
- self.dense2.bias: torch.Size([768])
where 8 is batch size, 1024 is sequence length.

**Comment by tridao** [comment]
Yes, I think for apex's fused_dense you'd need to combine the batch and seqlen dimensions.

**Comment by yongyanrao** [comment]
We can verify that when using apex's fused_dense along with Deepspeed, we did not notice any memory leaking issue. We tested it more than 1,000 steps.

**Comment by tridao** [comment]
Thanks, the memory issue with deepspeed might be because of the way I allocate temporary workspace. I'll try to fix in the next couple of days.

**Comment by yongyanrao** [comment]
Hi Tri, do we have any update on this front yet?

**Comment by tridao** [comment]
Can you try the latest version 1.0.7? I just changed sth on the workspace allocation for fused_dense.

**Comment by yongyanrao** [comment]
I tried v1.0.7, and the OOM issue still showed up.

**Comment by tridao** [comment]
Then I don't have any idea where the memory issue comes from.
